panther_analysis_tool/log_schemas/user_defined.py

```diff
@@ -0,0 +1,279 @@
+from dataclasses import dataclass
+import fnmatch
+import json
+import logging
+import os
+from typing import List, Dict, Tuple, Optional, Any, cast
+
+import boto3
+from botocore import client
+from ruamel.yaml import YAML
+from ruamel.yaml.scanner import ScannerError
+from ruamel.yaml.parser import ParserError
+
+
+logger = logging.getLogger(__file__)
+
+
+class Client:
+    _LAMBDA_NAME = 'panther-logtypes-api'
+    _LIST_SCHEMAS_ENDPOINT = 'ListSchemas'
+    _PUT_SCHEMA_ENDPOINT = 'PutUserSchema'
+
+    def __init__(self) -> None:
+        self._lambda_client = None
+
+    @property
+    def lambda_client(self) -> client.BaseClient:
+        if self._lambda_client is None:
+            self._lambda_client = boto3.client("lambda")
+        return self._lambda_client
+
+    def list_schemas(self) -> Tuple[bool, dict]:
+        return self._invoke(
+            self._create_lambda_request(
+                endpoint=self._LIST_SCHEMAS_ENDPOINT,
+                payload={
+                    'isManaged': False
+                }
+            )
+        )
+
+    def put_schema(self, name: str, definition: str, revision: int,  # pylint: disable=too-many-arguments
+                   description: str, reference_url: str) -> Tuple[bool, dict]:
```

Comment:
Having the revision be part of the request makes things more complicated dealing with this API. Not part of this PR but logtypes-api API codebase would be really simplified if we omitted that requirement.

Reply:
I think it was introduced to address some concurrency concerns, no? As a side note, there is also a great validation for detecting backwards incompatibilities, but I think it is based on the assumption that updates are performed sequentially (because it needs to compare with the previous version naturally). For such extra-low throughput / rarely called endpoints that perform critical updates, perhaps a pessimistic lock would be a good approach.